docs/markdown/snippets/asm.md

```diff
@@ -0,0 +1,17 @@
+## New language `nasm`
+
+When the `nasm` language is added to the project, `.asm` files are automatically
```

Comment:
I've never seen a project using `.nasm` extension, do you have examples in mind? At least vlc, ffmpeg and openssl all use `.asm` for nasm compiler.

Reply:
The goal here isn't to introduce a convention for people to add asm to new projects.

The goal is to fit into the convention that already exists.

The use of `.asm` is intentional and shall not change.

Reply:
True, I would like to follow what people actually do in real world projects out there, certainly not invent something new. If we have examples of people using `.nasm` then we can of course also associate that extension too.

Reply:
Here is one: https://github.com/tianocore/edk2/blob/master/OvmfPkg/Sec/X64/SecEntry.nasm. The problem is twofold:

* `.asm` can mean multiple things, it can mean both Microsoft and nasm syntaxes.  edk2 started to use `.nasm` because they wanted to use separate build rules for Microsoft and nasm syntaxes. Identifying build rules by the extension is something that Meson already does (e.g. anything ending with `*.c` _must_ be compiled with the C compiler).

* it's not even sure that assembly language it would require a separate `add_languages` on any architecture but x86.

There are other cases in which Meson says my way or the highway (e.g. no include files named `*.inc.c`), and projects that switch to Meson have to adapt (hello QEMU). I think in this case using `.nasm` is safer until OS-portable support for assemblers is fleshed out on other architectures. Adding stuff later is much easier than removing it.

Reply:
Ok, I could definitely map `.nasm ` to nasm language, that's unambiguous. I personally like the convention `.S` files goes through the C compiler, which can be either gcc/as or cl/ml, that seems unambiguous too.

But it's true that plain `.asm` can be ambiguous, ultimately it will have to be configurable I guess. How does edk2 compile their `.asm` currently?

> In the future we might want to allow `.asm` with MSVC just like we allow `.s` with GCC

Do you have examples of projects that does that? edk2? Maybe if we user does `add_language('nasm')` then `.asm` goes to nasm, otherwise if C compiler is MSVC then they go to cl/ml? Another solution that I had in mind for other non conventional C/C++ file extensions is adding an explicit API to map extension to language, such as `add_language('c', suffixes: ['.asm'])`.

Reply:
In the end here we only want to help the "common" case, it's fine to keep using custom_target()/generator() for more "exotic" conventions, that's already what they have to do right now anyway.

Reply:
> Do you have examples of projects that does that? edk2

Yes. Though by now I think they have converted everything to just nasm, they used to have some files as .S/.asm (synced by hand :)) and some as .nasm.

My suggestion is simply to not touch .asm with a ten foot pole for x86, until you're satisfied with the handling of .asm in other architectures. Assigning extensions in `add_languages` makes sense but again I would be wary of adding stuff that later will be unnecessary, or worse a poor design.

Reply:
>  I personally like the convention .S files goes through the C compiler, 

The problem is that .S on MSVC is unheard of. They use .asm. :(

Reply:
> The problem is that .S on MSVC is unheard of. They use .asm. :(

At least libffi does it:
```
[1/9] "C:\BuildTools\VC\Tools\MSVC\14.29.30037\bin\HostX64\x64\cl.EXE" "/nologo" "/EP" "/P" "/Fisrc/x86_win64_intel_S.i" "/DTARGET=X86_WIN64" "/DFFI_BUILDING" "../src/x86/win64_intel.S" "/IC:/Users/Collabora/xclaesse/libffi/src/.." "/IC:/Users/Collabora/xclaesse/libffi/build/src/.." "/IC:/Users/Collabora/xclaesse/libffi/src/../include" "/IC:/Users/Collabora/xclaesse/libffi/build/src/../include"
win64_intel.S
[7/9] "C:\BuildTools\VC\Tools\MSVC\14.29.30037\bin\HostX64\x64\ml64.EXE" "/Fo" "src/x86_win64_intel_S.obj" "/c" "src/x86_win64_intel_S.i" "/Zi" "/Zd"
Microsoft (R) Macro Assembler (x64) Version 14.29.30038.1
Copyright (C) Microsoft Corporation.  All rights reserved.

 Assembling: src/x86_win64_intel_S.i
```

Reply:
I too would be in favor of using `.nasm` to make it explicit.

Reply:
The question is not what we prefer, but what projects actually do. So far every projects I've seen need to build .asm with nasm. If a project has .asm files that are built with another compiler they just have to not add nasm language.